### PR TITLE
GSOC 2020: Migrate http / matrix implementations to futures 0.3

### PR DESCRIPTION
**Pull request by VanillaBrooks**
This PR swaps the http.rs implementation to use hyper for more readable and reliable http requests. Additionally, the matrix module is also updated to futures 0.3 so that async / await can be used in `bridge/mod.rs` in a future PR.

### CODE REVIEW

File: src/matrix/mod.rs [hunk 1]
```diff
@@ -278,7 +266,7 @@ impl JsonPostError {
 
 quick_error! {
     #[derive(Debug)]
-    pub enum LoginError {
+    pub enum Error {
```
Comment (VanillaBrooks) [inline src/matrix/mod.rs:269]: In order to remove the `.unwrap()`s from `join_room()` I unified all the error possibilities under one type. Previously this was done by many `.map_err()`s, and I thought this made things more readable. Thoughts on this?

File: src/main.rs [hunk 2]
```diff
@@ -20,8 +20,19 @@
...
+use slog_async;
+use slog_term;
+use tasked_futures;
+use tokio_core;
+use tokio_tls;
+use url;
+
 use clap::{App, Arg};
 
+use futures;
```
Comment (jplatte) [inline src/main.rs]: All these new imports are redundant. Why did you add them?

File: src/matrix/sync.rs [hunk 3]
```diff
@@ -12,33 +12,43 @@
...
     url: Url,
     access_token: String,
     next_token: Option<String>,
-    http_stream: HttpClient,
-    current_sync: Option<HttpResponseFuture>,
+    http_client: http::ClientWrapper,
+    current_sync: RequestStatus,
+}
+
+use std::boxed::Box;
```
Comment (jplatte) [inline src/matrix/sync.rs]: Why is this import all the way down here?

**Review by jplatte — COMMENTED** [review]
I have ignored the two largest file diffs for now, here's some feedback about the rest.

Also, CI failed with an error message about the lockfile being broken. It's fine if the code doesn't build because of types mismatches and such, but I didn't expect that. Did you do manual changes to the lock file?

File: src/main.rs (hunk 2 above)
Comment (VanillaBrooks) [reply src/main.rs]: These are issues that were solved in a previous PR, I have not rebased correctly so you are seeing ~5 (?) commits that have previously been pulled. Ill work on sorting this out.

**Comment by VanillaBrooks** [comment]
I have not touched the lock file at all. Every time I pull updates from matrix-org/matrix-ircd to my dev branch there is a merge conflict in the lock file (which previously I have never seen) which eventually leads to a cargo error. I opt to rebuild to lock file when this happens, is this the correct approach?

**Comment by jplatte** [comment]
Could you try rebasing instead of merging, and recreating the lockfile if there are merge conflicts or it fails with the same error at any step? If you need instructions, ping me on Matrix.

File: src/matrix/mod.rs (hunk 1 above)
Comment (phlmn) [reply src/matrix/mod.rs:269]: I think that's fine

File: src/matrix/sync.rs (hunk 3 above)
Comment (phlmn) [reply src/matrix/sync.rs]: @VanillaBrooks could you move this to the other imports?

**Review by phlmn — CHANGES_REQUESTED** [review]
I finally had time to look at the PR. Generally it looks good to me 🙂 . Happy to merge this as soon as  the findings @jplatte had are fixed!

**Comment by VanillaBrooks** [comment]
I think leaving the PR with build errors is the best thing to do here. The errors in `bridge/mod.rs` are fixed in the next commit on my async-await branch, but I unfortunately combined to commit with a bunch of IRC module changes. I also had trouble leaving `futures::compat` code in `bridge/mod.rs` to correctly build the project, but seeing as this is a development branch I think merging here is fine.

The changes to fix builder errors are [also very messy](https://github.com/VanillaBrooks/matrix-ircd/commit/d3dca604278894d88a7adc94e18d4a5a5b4942fb#diff-609a9a410b97b4f4ad7f14bd2a8f57ca), and depending on how we merge the next PR the mess might be introduced and immediately removed.
